src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Controllers/ResultsController.cs

```diff
@@ -21,59 +23,65 @@ public ResultsController(IResultsService resultsService)
-                            Week3Points = result.Week3Points,
-                            Week3Place = result.Week3Place,
-                            FinalPoints = result.FinalPoints,
-                            FinalPlace = result.FinalPlace,
-                            UserId = result.UserId,
-                            CorrectAnswers = result.CorrectAnswersCount,
-                            WrongAnswers = result.WrongAnswersCount,
-                            UserName = result.UserName,
-                        });
-                    }
-
-                    paginatedResults.Add(i, new PaginatedCollection<TestResultEntryViewModel>(totalTestResults, pageIndex ?? 1, pageSize));
-                }
-            }
-
             var userPosition = this._resultsService.GetUserPosition(userId);
             var vm = new TestResultsViewModel()
             {
                 UserWeek1Position = userPosition.Week1Place ?? 0,
                 UserWeek2Position = userPosition.Week2Place ?? 0,
                 UserWeek3Position = userPosition.Week3Place ?? 0,
-                UserFinalPosition = userPosition.FinalPlace ?? 0,
-                TotalTestResults = paginatedResults
+                UserFinalPosition = userPosition.FinalPlace ?? 0
             };
 
             return this.View(vm);
         }
+
+        [HttpGet]
+        public ActionResult RenderResults(int? pageIndex, int weekNumber = 1)
+        {
+            var userId = this.HttpContext.User.FindFirstValue(ClaimTypes.NameIdentifier);
+            if (userId == null)
+            {
+                return this.RedirectToAction("Index");
+            }
+
+            int pageSize = 50;
+
+            var testResultListDto = this._resultsService.GetTestResults(weekNumber);
+            if (testResultListDto == null || !testResultListDto.Any())
+            {
+                return this.PartialView("_ResultsPage", new KeyValuePair<int, PaginatedCollection<TestResultEntryViewModel>>(weekNumber, null));
+            }
+
+            var totalTestResults = testResultListDto
+                    .Select(t => new TestResultEntryViewModel
+                    {
+                        Week1Points = t.Week1Points,
+                        Week1Place = t.Week1Place,
+                        Week2Points = t.Week2Points,
+                        Week2Place = t.Week2Place,
+                        Week3Points = t.Week3Points,
+                        Week3Place = t.Week3Place,
+                        FinalPoints = t.FinalPoints,
+                        FinalPlace = t.FinalPlace,
+                        UserId = t.UserId,
+                        CorrectAnswers = t.CorrectAnswersCount,
+                        WrongAnswers = t.WrongAnswersCount,
+                        UserName = t.UserName,
+                    })
+                    .ToList();
+
+            var paginatedResults = new KeyValuePair<int, PaginatedCollection<TestResultEntryViewModel>>(
```

Comment:
we should load from db only 50 results, not all of them

Reply:
we are actually